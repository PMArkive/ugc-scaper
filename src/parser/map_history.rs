use super::Parser;
use crate::data::{CurrentSeasonMap, CurrentSeasonMapList, MapHistory};
use crate::parser::select_text;
use crate::{ParseError, Result};
use scraper::{Html, Selector};

const SELECTOR_CURRENT_ROW: &str = "table.table.table-condensed.table-responsive tbody tr";
const SELECTOR_CURRENT_SEASON: &str = "div.row > div > div.white-row-small > h5:nth-child(2), div.row-fluid > div > div.white-row-small > h4:first-child+h5";
const SELECTOR_CURRENT_WEEK: &str = "td:nth-child(1)";
const SELECTOR_CURRENT_MAP: &str = "td:nth-child(2)";
const SELECTOR_CURRENT_DATE: &str = "td:nth-child(4) small";
const SELECTOR_CURRENT_DATE_ALT: &str = "td:nth-child(5) small";

const SELECTOR_PREVIOUS: &str = "table.table.table-condensed.table-bordered";
const SELECTOR_PREVIOUS_SEASON: &str = "tr.top-bar td h3.text-info";

pub struct MapHistoryParser {
    selector_current_row: Selector,
    selector_current_season: Selector,
    selector_current_week: Selector,
    selector_current_map: Selector,
    selector_current_date: Selector,
    selector_current_date_alt: Selector,
}

impl Default for MapHistoryParser {
    fn default() -> Self {
        Self::new()
    }
}

impl MapHistoryParser {
    pub fn new() -> Self {
        MapHistoryParser {
            selector_current_row: Selector::parse(SELECTOR_CURRENT_ROW).unwrap(),
            selector_current_season: Selector::parse(SELECTOR_CURRENT_SEASON).unwrap(),
            selector_current_week: Selector::parse(SELECTOR_CURRENT_WEEK).unwrap(),
            selector_current_map: Selector::parse(SELECTOR_CURRENT_MAP).unwrap(),
            selector_current_date: Selector::parse(SELECTOR_CURRENT_DATE).unwrap(),
            selector_current_date_alt: Selector::parse(SELECTOR_CURRENT_DATE_ALT).unwrap(),
        }
    }
}

impl Parser for MapHistoryParser {
    type Output = MapHistory;

    fn parse(&self, document: &str) -> Result<Self::Output> {
        let document = Html::parse_document(document);

        let season = select_text(document.root_element(), &self.selector_current_season)
            .ok_or(ParseError::ElementNotFound {
                selector: SELECTOR_CURRENT_SEASON,
                role: "current season number",
            })?
            .trim_start_matches("Season")
            .trim();
        let season: u8 = season.parse().map_err(|_| ParseError::InvalidText {
            role: "current season number",
            text: season.to_string(),
        })?;

        let current_weeks = document
            .select(&self.selector_current_row)
            .map(|row| {
                let week = select_text(row, &self.selector_current_week).ok_or(
                    ParseError::ElementNotFound {
                        selector: SELECTOR_CURRENT_WEEK,
                        role: "current season week number",
                    },
                )?;
                let week: u8 = week.parse().map_err(|_| ParseError::InvalidText {
                    role: "current season week number",
                    text: season.to_string(),
                })?;
                let map = select_text(row, &self.selector_current_map)
                    .ok_or(ParseError::ElementNotFound {
                        selector: SELECTOR_CURRENT_MAP,
                        role: "current season map",
                    })?
                    .to_string();
                let mut date = select_text(row, &self.selector_current_date)
                    .ok_or(ParseError::ElementNotFound {
                        selector: SELECTOR_CURRENT_MAP,
                        role: "current season map",
                    })?
                    .to_string();
                let mut alt_date =
                    select_text(row, &self.selector_current_date_alt).map(String::from);
                if let Some(global_date) = alt_date.take() {
                    alt_date = Some(date);
                    date = global_date;
                }

                Ok(CurrentSeasonMap {
                    week,
                    map,
                    date,
                    na_date: alt_date,
                })
            })
            .collect::<Result<_>>()?;

        Ok(MapHistory {
            current: CurrentSeasonMapList {
                season,
                maps: current_weeks,
            },
            previous: Vec::new(),
        })
    }
}
