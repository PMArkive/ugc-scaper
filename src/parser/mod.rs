use crate::{ParseError, Result};
use scraper::{ElementRef, Selector};
use steamid_ng::SteamID;
use time::format_description::FormatItem;
use time::macros::format_description;

mod player;
mod player_details;
mod seasons;
mod team;
mod team_lookup;
mod team_matches;
mod team_roster_history;

pub use player::*;
pub use player_details::*;
pub use seasons::*;
pub use team::*;
pub use team_lookup::*;
pub use team_matches::*;
pub use team_roster_history::*;

pub trait Parser {
    type Output;
    fn parse(&self, document: &str) -> Result<Self::Output>;
}

trait ElementExt<'a> {
    fn first_text(&'a self) -> Option<&'a str>;
    fn nth_text(&'a self, n: usize) -> Option<&'a str>;
}

impl<'a> ElementExt<'a> for ElementRef<'a> {
    fn first_text(&self) -> Option<&'a str> {
        self.text().map(str::trim).find(|s| !s.is_empty())
    }
    fn nth_text(&self, n: usize) -> Option<&'a str> {
        self.text()
            .filter(|s| !s.trim().is_empty())
            .nth(n - 1)
            .map(str::trim)
    }
}

fn select_text<'a>(el: ElementRef<'a>, selector: &Selector) -> Option<&'a str> {
    el.select(selector)
        .next()
        .and_then(|item| item.text().find(|s| !s.trim().is_empty()))
        .map(str::trim)
}

fn select_last_text<'a>(el: ElementRef<'a>, selector: &Selector) -> Option<&'a str> {
    el.select(selector)
        .next()
        .and_then(|item| item.text().last())
        .map(str::trim)
}

const DATE_FORMAT: &[FormatItem<'static>] =
    format_description!("[month padding:none]/[day padding:none]/[year]");
const MEMBER_DATE_FORMAT: &[FormatItem<'static>] = format_description!(
    "[month repr:short] [day padding:none], [year]\n/\n[hour padding:none]:[minute] [period]\n(ET)"
);
const ROSTER_HISTORY_DATE_FORMAT: &[FormatItem<'static>] =
    format_description!("[month repr:short] [day padding:none], [year]");

fn team_id_from_link(link: &str) -> Result<u32, ParseError> {
    link.rsplit_once('=')
        .and_then(|part| part.1.parse().ok())
        .ok_or_else(|| ParseError::InvalidLink {
            link: link.to_string(),
            role: "team id",
        })
}

fn steam_id_from_link(link: &str) -> Result<SteamID, ParseError> {
    link.rsplit_once('=')
        .and_then(|part| part.1.parse::<u64>().ok())
        .ok_or_else(|| ParseError::InvalidLink {
            link: link.to_string(),
            role: "user id",
        })
        .map(SteamID::from)
}
