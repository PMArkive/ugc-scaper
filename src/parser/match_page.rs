use super::Parser;
use crate::data::{GameMode, MatchInfo, TeamRef};
use crate::parser::{select_last_text, select_text, team_id_from_link, ElementExt};
use crate::{ParseError, Result};
use scraper::{Html, Selector};

const SELECTOR_MATCH_FORMAT: &str = "h3.page-header > strong.styleColor";
const SELECTOR_MATCH_COMMENT_AUTHOR: &str = ".row-fluid .col-md-12 span.text-success";
const SELECTOR_MATCH_COMMENT: &str = ".row-fluid .col-md-12 > .white-row-light-small > p";
const SELECTOR_MATCH_TEAM_LINK: &str = "a[href^=\"team_page\"]:not(.btn-large)";
const SELECTOR_MATCH_RESULT_TEAM: &str =
    ".table.table-condensed.table-bordered tr:nth-child(2) td:nth-child(1)";
const SELECTOR_MATCH_RESULT_SCORE: &str =
    ".table.table-condensed.table-bordered tr:nth-child(2) td:nth-child(2)";
const SELECTOR_MATCH_MAP: &str = "h4.text-success.text-center > b";
const SELECTOR_MATCH_WEEK: &str = "p.muted.text-center.nomargin > small > b:nth-child(1)";
const SELECTOR_MATCH_DATE: &str = "p.muted.text-center.nomargin > small > b:nth-child(2)";

pub struct MatchPageParser {
    selector_format: Selector,
    selector_author: Selector,
    selector_comment: Selector,
    selector_team_link: Selector,
    selector_result_team: Selector,
    selector_result_score: Selector,
    selector_map: Selector,
    selector_week: Selector,
    selector_date: Selector,
}

impl Default for MatchPageParser {
    fn default() -> Self {
        Self::new()
    }
}

impl MatchPageParser {
    pub fn new() -> Self {
        MatchPageParser {
            selector_format: Selector::parse(SELECTOR_MATCH_FORMAT).unwrap(),
            selector_author: Selector::parse(SELECTOR_MATCH_COMMENT_AUTHOR).unwrap(),
            selector_comment: Selector::parse(SELECTOR_MATCH_COMMENT).unwrap(),
            selector_team_link: Selector::parse(SELECTOR_MATCH_TEAM_LINK).unwrap(),
            selector_result_team: Selector::parse(SELECTOR_MATCH_RESULT_TEAM).unwrap(),
            selector_result_score: Selector::parse(SELECTOR_MATCH_RESULT_SCORE).unwrap(),
            selector_map: Selector::parse(SELECTOR_MATCH_MAP).unwrap(),
            selector_week: Selector::parse(SELECTOR_MATCH_WEEK).unwrap(),
            selector_date: Selector::parse(SELECTOR_MATCH_DATE).unwrap(),
        }
    }
}

impl Parser for MatchPageParser {
    type Output = MatchInfo;

    fn parse(&self, document: &str) -> Result<Self::Output> {
        let document = Html::parse_document(document);

        let author = select_text(document.root_element(), &self.selector_author);
        let comment = select_last_text(document.root_element(), &self.selector_comment);

        let mut team_links = document.select(&self.selector_team_link);
        let team_link_home = team_links.next().ok_or(ParseError::ElementNotFound {
            selector: SELECTOR_MATCH_TEAM_LINK,
            role: "home team link",
        })?;
        let team_link_away = team_links.next().ok_or(ParseError::ElementNotFound {
            selector: SELECTOR_MATCH_TEAM_LINK,
            role: "away team link",
        })?;
        let home_team_id = team_id_from_link(team_link_home.attr("href").unwrap_or_default())?;
        let away_team_id = team_id_from_link(team_link_away.attr("href").unwrap_or_default())?;

        let format_text = select_text(document.root_element(), &self.selector_format).ok_or(
            ParseError::ElementNotFound {
                selector: SELECTOR_MATCH_FORMAT,
                role: "away team map",
            },
        )?;
        let format = format_text
            .parse::<GameMode>()
            .ok()
            .or_else(|| format_text.split(' ').find_map(|text| text.parse().ok()))
            .ok_or_else(|| ParseError::InvalidText {
                text: format_text.into(),
                role: "match format",
            })?;

        let map = select_text(document.root_element(), &self.selector_map).ok_or(
            ParseError::ElementNotFound {
                selector: SELECTOR_MATCH_MAP,
                role: "away team map",
            },
        )?;

        let week_text = select_text(document.root_element(), &self.selector_week).ok_or(
            ParseError::ElementNotFound {
                selector: SELECTOR_MATCH_WEEK,
                role: "away team week",
            },
        )?;
        let week: u8 = week_text.parse().map_err(|_| ParseError::InvalidText {
            text: week_text.into(),
            role: "match week",
        })?;

        let date = select_text(document.root_element(), &self.selector_date).ok_or(
            ParseError::ElementNotFound {
                selector: SELECTOR_MATCH_DATE,
                role: "away team week",
            },
        )?;

        let mut team_names = document.select(&self.selector_result_team);
        let team_name_home = team_names
            .next()
            .ok_or(ParseError::ElementNotFound {
                selector: SELECTOR_MATCH_RESULT_TEAM,
                role: "home team link",
            })?
            .first_text()
            .ok_or(ParseError::EmptyText {
                role: "home team name",
                selector: SELECTOR_MATCH_RESULT_TEAM,
            })?
            .to_string();
        let team_name_away = team_names
            .next()
            .ok_or(ParseError::ElementNotFound {
                selector: SELECTOR_MATCH_RESULT_TEAM,
                role: "away team link",
            })?
            .first_text()
            .ok_or(ParseError::EmptyText {
                role: "away team name",
                selector: SELECTOR_MATCH_RESULT_TEAM,
            })?
            .to_string();

        let mut team_scores = document.select(&self.selector_result_score);

        let team_score_home = team_scores
            .next()
            .ok_or(ParseError::ElementNotFound {
                selector: SELECTOR_MATCH_RESULT_SCORE,
                role: "home team score",
            })?
            .first_text()
            .ok_or(ParseError::EmptyText {
                role: "home team score",
                selector: SELECTOR_MATCH_RESULT_SCORE,
            })?
            .parse()
            .map_err(|_| ParseError::InvalidText {
                role: "away team score",
                text: "dont have this".to_string(),
            })?;
        let team_score_away = team_scores
            .next()
            .ok_or(ParseError::ElementNotFound {
                selector: SELECTOR_MATCH_RESULT_SCORE,
                role: "away team link",
            })?
            .first_text()
            .ok_or(ParseError::EmptyText {
                role: "away team name",
                selector: SELECTOR_MATCH_RESULT_SCORE,
            })?
            .parse()
            .map_err(|_| ParseError::InvalidText {
                role: "home team score",
                text: "dont have this".to_string(),
            })?;

        Ok(MatchInfo {
            comment_author: author.map(String::from),
            comment: comment.map(String::from),
            score_away: team_score_away,
            score_home: team_score_home,
            team_home: TeamRef {
                name: team_name_home.to_string(),
                id: home_team_id,
            },
            team_away: TeamRef {
                name: team_name_away.to_string(),
                id: away_team_id,
            },
            week,
            map: map.into(),
            default_date: date.into(),
            format,
        })
    }
}
