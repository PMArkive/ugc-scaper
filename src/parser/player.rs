use super::{ElementExt, Parser};
use crate::data::{Class, Honors, Player, TeamMemberShip, TeamRef};
use crate::parser::{select_last_text, select_text, team_id_from_link, DATE_FORMAT};
use crate::{ParseError, Result};
use scraper::{Html, Selector};
use std::iter::repeat;
use std::str::FromStr;
use steamid_ng::SteamID;
use time::Date;

const SELECTOR_PLAYER_NAME: &str = ".container .col-md-4 > h3 > b";
const SELECTOR_PLAYER_ID: &str = ".container .col-md-4 > p.nomargin";

const SELECTOR_PLAYER_HONORS_GROUP: &str =
    ".container .col-md-6:nth-child(2) .white-row-small .row-fluid";
const SELECTOR_PLAYER_HONORS_HEADER: &str = "h5";
const SELECTOR_PLAYER_HONORS_LEAGUE: &str = "li div";
const SELECTOR_PLAYER_HONORS_TEAM: &str = "li small";

const SELECTOR_PLAYER_TEAM_GROUP: &str =
    ".container .col-md-6:nth-child(1) .white-row-small .row-fluid";
const SELECTOR_PLAYER_TEAM_LINK: &str = "p a";
const SELECTOR_PLAYER_TEAM_NAME: &str = "span.text-primary b";
const SELECTOR_PLAYER_TEAM_LEAGUE: &str = "small";
const SELECTOR_PLAYER_TEAM_SINCE: &str = "small";

const SELECTOR_CLASS: &str = r#"img.img-rounded[src*="images/tf2/icon/"]"#;

pub struct PlayerParser {
    selector_name: Selector,
    selector_id: Selector,

    selector_honors_header: Selector,
    selector_honors_group: Selector,
    selector_honors_league: Selector,
    selector_honors_team: Selector,

    selector_team_group: Selector,
    selector_team_link: Selector,
    selector_team_name: Selector,
    selector_team_league: Selector,
    selector_team_since: Selector,

    selector_class: Selector,
}

impl Default for PlayerParser {
    fn default() -> Self {
        Self::new()
    }
}

impl PlayerParser {
    pub fn new() -> Self {
        PlayerParser {
            selector_name: Selector::parse(SELECTOR_PLAYER_NAME).unwrap(),
            selector_id: Selector::parse(SELECTOR_PLAYER_ID).unwrap(),

            selector_honors_header: Selector::parse(SELECTOR_PLAYER_HONORS_HEADER).unwrap(),
            selector_honors_group: Selector::parse(SELECTOR_PLAYER_HONORS_GROUP).unwrap(),
            selector_honors_league: Selector::parse(SELECTOR_PLAYER_HONORS_LEAGUE).unwrap(),
            selector_honors_team: Selector::parse(SELECTOR_PLAYER_HONORS_TEAM).unwrap(),

            selector_team_group: Selector::parse(SELECTOR_PLAYER_TEAM_GROUP).unwrap(),
            selector_team_link: Selector::parse(SELECTOR_PLAYER_TEAM_LINK).unwrap(),
            selector_team_name: Selector::parse(SELECTOR_PLAYER_TEAM_NAME).unwrap(),
            selector_team_league: Selector::parse(SELECTOR_PLAYER_TEAM_LEAGUE).unwrap(),
            selector_team_since: Selector::parse(SELECTOR_PLAYER_TEAM_SINCE).unwrap(),

            selector_class: Selector::parse(SELECTOR_CLASS).unwrap(),
        }
    }
}

impl Parser for PlayerParser {
    type Output = Player;

    fn parse(&self, document: &str) -> Result<Self::Output> {
        let document = Html::parse_document(document);
        let name = document
            .select(&self.selector_name)
            .next()
            .ok_or(ParseError::ElementNotFound {
                selector: SELECTOR_PLAYER_NAME,
                role: "player name",
            })?
            .first_text()
            .unwrap_or_default()
            .to_string();

        let id = document
            .select(&self.selector_id)
            .next()
            .ok_or(ParseError::ElementNotFound {
                selector: SELECTOR_PLAYER_ID,
                role: "player steam id",
            })?
            .nth_text(3)
            .unwrap_or_default()
            .to_string();

        let honors = document
            .select(&self.selector_honors_group)
            .flat_map(|group| {
                let format = select_text(group, &self.selector_honors_header)
                    .ok_or(ParseError::ElementNotFound {
                        selector: SELECTOR_PLAYER_HONORS_HEADER,
                        role: "player honors format",
                    })
                    .map(|format| format.trim_end_matches(" Medals"));
                let leagues = group.select(&self.selector_honors_league);
                let teams = group.select(&self.selector_honors_team);
                repeat(format).zip(leagues).zip(teams)
            })
            .map(|((format_res, season), team)| {
                let format = format_res?;
                Ok(Honors {
                    format: format.to_string(),
                    season: season
                        .first_text()
                        .ok_or(ParseError::EmptyText {
                            selector: SELECTOR_PLAYER_HONORS_LEAGUE,
                            role: "player honors season",
                        })?
                        .to_string(),
                    team: team
                        .first_text()
                        .ok_or(ParseError::EmptyText {
                            selector: SELECTOR_PLAYER_HONORS_TEAM,
                            role: "player honors team",
                        })?
                        .to_string(),
                })
            })
            .collect::<Result<Vec<_>>>()?;

        let favorite_classes = document
            .select(&self.selector_class)
            .filter_map(|class| class.attr("src"))
            .filter_map(|img| {
                img.strip_prefix("images/tf2/icon/")
                    .and_then(|class| class.strip_suffix(".jpg"))
            })
            .filter_map(|class| Class::from_str(class).ok())
            .collect::<Vec<_>>();

        let teams = document
            .select(&self.selector_team_group)
            .filter(|item| item.select(&self.selector_team_link).next().is_some())
            .map(move |item| {
                let link = item
                    .select(&self.selector_team_link)
                    .next()
                    .ok_or(ParseError::ElementNotFound {
                        selector: SELECTOR_PLAYER_TEAM_LINK,
                        role: "players team link",
                    })?
                    .attr("href")
                    .unwrap_or_default();
                let name = select_text(item, &self.selector_team_name).ok_or(
                    ParseError::ElementNotFound {
                        selector: SELECTOR_PLAYER_TEAM_NAME,
                        role: "players team name",
                    },
                )?;
                let league = select_text(item, &self.selector_team_league).ok_or(
                    ParseError::ElementNotFound {
                        selector: SELECTOR_PLAYER_TEAM_LEAGUE,
                        role: "players team league",
                    },
                )?;
                let since = select_last_text(item, &self.selector_team_since).ok_or(
                    ParseError::ElementNotFound {
                        selector: SELECTOR_PLAYER_TEAM_SINCE,
                        role: "players team joined",
                    },
                )?;

                let id = team_id_from_link(link)?;
                let since = match since.rsplit_once('\n') {
                    Some((_, since)) => {
                        Date::parse(since, DATE_FORMAT).map_err(|_| ParseError::InvalidDate {
                            role: "team join date",
                            date: since.to_string(),
                        })?
                    }
                    _ => {
                        return Err(ParseError::InvalidDate {
                            role: "team join date",
                            date: since.to_string(),
                        }
                        .into())
                    }
                };

                Ok(TeamMemberShip {
                    team: TeamRef {
                        name: name.to_string(),
                        id,
                    },
                    league: league.to_string(),
                    since,
                })
            })
            .collect::<Result<Vec<_>>>()?;

        Ok(Player {
            name,
            steam_id: SteamID::from_steam3(&id).unwrap_or_default(),
            honors,
            teams,
            favorite_classes,
        })
    }
}
