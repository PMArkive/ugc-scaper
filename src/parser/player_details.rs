use super::{ElementExt, Parser};
use crate::data::{MembershipHistory, TeamRef};
use crate::parser::{select_text, team_id_from_link, DATE_FORMAT};
use crate::{ParseError, Result};
use scraper::{Html, Selector};
use time::Date;

const SELECTOR_TEAM_FORMAT: &str = ".container .white-row-small thead h4";
const SELECTOR_TEAM_GROUP: &str = ".container .white-row-small tbody";
const TEAM_ROW: &str = "tr:not(:first-child)";
const SELECTOR_TEAM_LINK: &str = "td:nth-child(3) a";
const SELECTOR_TEAM_DIVISION: &str = "td:nth-child(3) small";
const SELECTOR_TEAM_JOINED: &str = "td:nth-child(5) span";
const SELECTOR_TEAM_LEFT: &str = "td:nth-child(6) span";

pub struct PlayerDetailsParser {
    selector_team_format: Selector,
    selector_team_group: Selector,
    selector_team_row: Selector,
    selector_team_link: Selector,
    selector_team_division: Selector,
    selector_team_joined: Selector,
    selector_team_left: Selector,
}

impl PlayerDetailsParser {
    pub fn new() -> Self {
        PlayerDetailsParser {
            selector_team_format: Selector::parse(SELECTOR_TEAM_FORMAT).unwrap(),
            selector_team_group: Selector::parse(SELECTOR_TEAM_GROUP).unwrap(),
            selector_team_row: Selector::parse(TEAM_ROW).unwrap(),
            selector_team_link: Selector::parse(SELECTOR_TEAM_LINK).unwrap(),
            selector_team_division: Selector::parse(SELECTOR_TEAM_DIVISION).unwrap(),
            selector_team_joined: Selector::parse(SELECTOR_TEAM_JOINED).unwrap(),
            selector_team_left: Selector::parse(SELECTOR_TEAM_LEFT).unwrap(),
        }
    }
}

impl Default for PlayerDetailsParser {
    fn default() -> Self {
        Self::new()
    }
}

impl Parser for PlayerDetailsParser {
    type Output = Vec<MembershipHistory>;

    fn parse(&self, document: &str) -> Result<Self::Output> {
        let document = Html::parse_document(document);

        document
            .select(&self.selector_team_format)
            .zip(document.select(&self.selector_team_group))
            .flat_map(|(format, history)| {
                history
                    .select(&self.selector_team_row)
                    .map(move |row| (format, row))
            })
            .map(|(format, team)| {
                let format = format.first_text().ok_or(ParseError::EmptyText {
                    selector: SELECTOR_TEAM_FORMAT,
                    role: "team format",
                })?;
                let link = team
                    .select(&self.selector_team_link)
                    .next()
                    .ok_or(ParseError::ElementNotFound {
                        selector: SELECTOR_TEAM_LINK,
                        role: "team link",
                    })?
                    .attr("href")
                    .unwrap_or_default();
                let name = select_text(team, &self.selector_team_link).ok_or(
                    ParseError::ElementNotFound {
                        selector: SELECTOR_TEAM_LINK,
                        role: "team link",
                    },
                )?;
                let division = select_text(team, &self.selector_team_division).ok_or(
                    ParseError::ElementNotFound {
                        selector: SELECTOR_TEAM_DIVISION,
                        role: "team division",
                    },
                )?;
                let joined = select_text(team, &self.selector_team_joined).ok_or(
                    ParseError::ElementNotFound {
                        selector: SELECTOR_TEAM_JOINED,
                        role: "team join date",
                    },
                )?;
                let left = select_text(team, &self.selector_team_left).unwrap_or_default();

                let id = team_id_from_link(link)?;

                Ok(MembershipHistory {
                    format: format.to_string(),
                    joined: Date::parse(joined, DATE_FORMAT).map_err(|_| {
                        ParseError::InvalidDate {
                            role: "team join date",
                            date: joined.to_string(),
                        }
                    })?,
                    left: Date::parse(left, DATE_FORMAT).ok(),
                    team: TeamRef {
                        name: name.to_string(),
                        id,
                    },
                    division: division.to_string(),
                })
            })
            .collect()
    }
}
