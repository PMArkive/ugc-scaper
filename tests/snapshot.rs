#[cfg(feature = "serde")]
use insta::assert_json_snapshot;
#[cfg(feature = "serde")]
use std::fs::read_to_string;
#[cfg(feature = "serde")]
use test_case::test_case;
#[cfg(feature = "serde")]
use ugc_scraper::parser::{
    MapHistoryParser, MatchPageParser, Parser, PlayerDetailsParser, PlayerParser, SeasonsParser,
    TeamLookupParser, TeamMatchesParser, TeamParser, TeamRosterHistoryParser, TransactionParser,
};

#[test_case("player_76561198024494988.html", "player")]
#[test_case("player_76561198049312442.html", "player_classes")]
#[cfg(feature = "serde")]
fn test_parse_player_html(input: &str, name: &str) {
    let body = read_to_string(format!("tests/data/{input}")).unwrap();
    let parser = PlayerParser::new();
    let parsed = parser.parse(&body).unwrap();
    assert_json_snapshot!(format!("parse_{name}_html"), parsed);
}

#[test_case("player_details_76561198024494988.html", "player_details")]
#[cfg(feature = "serde")]
fn test_parse_player_details_html(input: &str, name: &str) {
    let body = read_to_string(format!("tests/data/{input}")).unwrap();
    let parser = PlayerDetailsParser::new();
    let parsed = parser.parse(&body).unwrap();
    assert_json_snapshot!(format!("parse_{name}_html"), parsed);
}

#[test_case("team_7861.html", "team")]
#[test_case("team_4105.html", "older_team")]
#[test_case("team_8157.html", "team_no_tz")]
#[test_case("team_6929.html", "team_changed_name")]
#[test_case("team_32437.html", "team_empty_name_change")]
#[test_case("team_29228.html", "team_newlines_join_date")]
#[test_case("team_10763.html", "team_na_4v4")]
#[cfg(feature = "serde")]
fn test_parse_team_html(input: &str, name: &str) {
    let body = read_to_string(format!("tests/data/{input}")).unwrap();
    let parser = TeamParser::new();
    let parsed = parser.parse(&body).unwrap();
    assert_json_snapshot!(format!("parse_{name}_html"), parsed);
}

#[test_case("team_roster_history_7861.html", "team_roster_history")]
#[test_case("team_roster_history_2133.html", "team_roster_history_no_group")]
#[cfg(feature = "serde")]
fn test_parse_team_roster_history_html(input: &str, name: &str) {
    let body = read_to_string(format!("tests/data/{input}")).unwrap();
    let parser = TeamRosterHistoryParser::new();
    let parsed = parser.parse(&body).unwrap();
    assert_json_snapshot!(format!("parse_{name}_html"), parsed);
}

#[test_case("team_matches_7861.html", "team_matches")]
#[cfg(feature = "serde")]
fn test_parse_team_matches_html(input: &str, name: &str) {
    let body = read_to_string(format!("tests/data/{input}")).unwrap();
    let parser = TeamMatchesParser::new();
    let parsed = parser.parse(&body).unwrap();
    assert_json_snapshot!(format!("parse_{name}_html"), parsed);
}

#[test_case("index.html", "seasons")]
#[cfg(feature = "serde")]
fn test_parse_seasons_html(input: &str, name: &str) {
    let body = read_to_string(format!("tests/data/{input}")).unwrap();
    let parser = SeasonsParser::new();
    let parsed = parser.parse(&body).unwrap();
    assert_json_snapshot!(format!("parse_{name}_html"), parsed);
}

#[test_case("lookup_9v9.html", "seasons_9")]
#[test_case("lookup_6v6.html", "seasons_6")]
#[test_case("lookup_4v4.html", "seasons_4")]
#[test_case("lookup_2v2.html", "seasons_2")]
#[cfg(feature = "serde")]
fn test_parse_seasons_mode_html(input: &str, name: &str) {
    let body = read_to_string(format!("tests/data/{input}")).unwrap();
    let parser = TeamLookupParser::new();
    let parsed = parser.parse(&body).unwrap();
    assert_json_snapshot!(format!("parse_{name}_html"), parsed);
}

#[test_case("match_116246.html", "match")]
#[cfg(feature = "serde")]
fn test_parse_match_html(input: &str, name: &str) {
    let body = read_to_string(format!("tests/data/{input}")).unwrap();
    let parser = MatchPageParser::new();
    let parsed = parser.parse(&body).unwrap();
    assert_json_snapshot!(format!("parse_{name}_html"), parsed);
}

#[test_case("transactions_4v4.html", "transaction")]
#[cfg(feature = "serde")]
fn test_parse_transaction_html(input: &str, name: &str) {
    let body = read_to_string(format!("tests/data/{input}")).unwrap();
    let parser = TransactionParser::new();
    let parsed = parser.parse(&body).unwrap();
    assert_json_snapshot!(format!("parse_{name}_html"), parsed);
}

#[test_case("map_9v9.html", "maps_9")]
#[test_case("map_9v9_2.html", "maps_9_2")]
#[test_case("map_6v6.html", "maps_6")]
#[test_case("map_4v4.html", "maps_4")]
#[test_case("map_2v2.html", "maps_2")]
#[cfg(feature = "serde")]
fn test_parse_maps_html(input: &str, name: &str) {
    let body = read_to_string(format!("tests/data/{input}")).unwrap();
    let parser = MapHistoryParser::new();
    let parsed = parser.parse(&body).unwrap();
    assert_json_snapshot!(format!("parse_{name}_html"), parsed);
}
