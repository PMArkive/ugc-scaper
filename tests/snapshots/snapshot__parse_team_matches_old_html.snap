---
source: tests/snapshot.rs
expression: parsed
---
{
  "team": {
    "name": "ChaosTheory",
    "id": 2157
  },
  "seasons": [
    {
      "season": 1,
      "format": "8v8",
      "matches": [
        {
          "division": "Steel",
          "week": 1,
          "date": "Sun Oct 28",
          "side": "visiting",
          "result": {
            "state": "unknown",
            "opponent": {
              "name": "[$0.00]",
              "id": 2137
            }
          },
          "map": "cp_gravelpit"
        },
        {
          "division": "Platinum",
          "week": 2,
          "date": "Sun Nov 04",
          "side": "home",
          "result": {
            "state": "unknown",
            "opponent": {
              "name": "-DeX-",
              "id": 2130
            }
          },
          "map": "cp_gravelpit"
        },
        {
          "division": "Platinum",
          "week": 3,
          "date": "Sun Nov 11",
          "side": "home",
          "result": {
            "state": "unknown",
            "opponent": {
              "name": "[UGS]",
              "id": 2127
            }
          },
          "map": "cp_dustbowl"
        },
        {
          "division": "Platinum",
          "week": 4,
          "date": "Sun Nov 18",
          "side": "home",
          "result": {
            "state": "bye_week"
          },
          "map": "cp_dustbowl"
        },
        {
          "division": "Platinum",
          "week": 6,
          "date": "Mon Dec 10",
          "side": "home",
          "result": {
            "state": "unknown",
            "opponent": {
              "name": ".tKd",
              "id": 2169
            }
          },
          "map": "ctf_impact"
        },
        {
          "division": "Platinum",
          "week": 7,
          "date": "Mon Dec 17",
          "side": "visiting",
          "result": {
            "state": "unknown",
            "opponent": {
              "name": "[FOoM]",
              "id": 2172
            }
          },
          "map": "cp_science2"
        },
        {
          "division": "Platinum",
          "week": 11,
          "date": "Mon Jan 14",
          "side": "home",
          "result": {
            "state": "unknown",
            "opponent": {
              "name": "[SWAT]",
              "id": 2188
            }
          },
          "map": "cp_junction"
        },
        {
          "division": "Platinum",
          "week": 12,
          "date": "Mon Jan 21",
          "side": "home",
          "result": {
            "state": "unknown",
            "opponent": {
              "name": "[FOoM]",
              "id": 2172
            }
          },
          "map": "cp_castle3"
        },
        {
          "division": "Platinum",
          "week": 13,
          "date": "Mon Jan 28",
          "side": "home",
          "result": {
            "state": "unknown",
            "opponent": {
              "name": "[=1=]",
              "id": 2197
            }
          },
          "map": "cp_dustbowl"
        },
        {
          "division": "Platinum",
          "week": 14,
          "date": "Mon Feb 04",
          "side": "home",
          "result": {
            "state": "unknown",
            "opponent": {
              "name": "[FOoM]",
              "id": 2172
            }
          },
          "map": "cp_science"
        }
      ]
    }
  ]
}
